docs/proposals/201812_thanos-remote-receive.md

```diff
@@ -0,0 +1,166 @@
+                                 |  +-----------------+  |        |
+                                 |  |                 |  |        |
+                                 |  | Thanos receiver +<----------+
+                                 |  |                 |  |        |
+                                 |  +-----------------+  |        |
+                                 |                       |        |
+                                 |                       |        |
+                                 |  +-----------------+  |        |
+                                 |  |                 |  |        |
+                                 |  | Thanos receiver +<----------+
+                                 |  |                 |  |
+                                 |  +-----------------+  |
+                                 |                       |
+                                 +-----------------------+
+```
+
+The intention is that the load balancer can just distribute requests randomly to all Thanos receivers independent of the tenant. Should this turn out to cause problems, either an additional “distribution” layer can be inserted that performs this routing or a load balancer per hard tenant hashring can be introduced. The distribution layer would have the advantage of being able to keep a single endpoint for configuration of Prometheus servers for remote write, which could be convenient for users.
+
+### Rollout/scaling/failure of receiver nodes
+
+As replication is based on the Prometheus write-ahead-log and retries when the remote write backend is not available, intermediate downtime is tolerable and expected for receivers. On rollouts receivers do not need to re-shard data, but instead at shutdown in case of rollout or scaling flush the write-ahead-log to a Prometheus tsdb block and upload it to object storage. Rollouts that include a soft tenant being promoted to a hard tenant, does require all nodes of a hash-ring to upload its content as the hash-ring changes. When the nodes comes back and accepts remote write requests again, the tenant local Prometheus server will continue where it left off. When scaling, all nodes need to perform the above operation as the hashring is resized meaning all nodes will have a new distribution. In the case of a failure, and the hashring is not resized, it will load the write-ahead-log and assume where it left off. Partially succeeding requests return a 503 causing Prometheus to retry the full request. This works as identically existing timestamp-value matches are ignored by tsdb. Prometheus relies on this to de-duplicate federation request results, therefore it is safe to rely on this here as well.
+
+This may produce small, and therefore unoptimized, blocks of TSDB in object storage, however, these are optimized away, by the Thanos compactor by merging the small blocks into bigger blocks. The compaction process is done concurrently in a separate deployment to the receivers.
+
```

Comment:
Will clock drift effect this at all ... if a node comes up and its clock is behind we may end up with overlapping blocks ... I suppose this is something we should fix in compactor when we see it.

Reply:
Timestamps are still produced by the originally ingesting Prometheus server, so while clock drift might happen on receivers, it doesn't effect any of the data stored in the TSDBs.